File: DataSets/DataSet-New/LeetCodeTasks - Develop/LeetCodeTasks/3083_existence-of-a-substring-in-a-string-and-its-reverse/solutions/cpp/solution.cpp
#include <string>
#include <unordered_set>
using namespace std;

class Solution {
public:
    bool isSubstringPresent(string s) {
        unordered_set<string> substrings;
        int n = s.length();
        
        // Collect all substrings of length 2 in the original string
        for (int i = 0; i < n - 1; ++i) {
            substrings.insert(s.substr(i, 2));
        }
        
        // Check if any substring from the original is in the reversed string
        for (int i = n - 1; i > 0; --i) {
            string revSubstr = s.substr(i - 1, 2);
            if (substrings.count(revSubstr)) {
                return true;
            }
        }
        
        return false;
    }
};

File: DataSets/DataSet-New/LeetCodeTasks - Develop/LeetCodeTasks/3083_existence-of-a-substring-in-a-string-and-its-reverse/solutions/java/solution.java
class Solution {
    public boolean isSubstringPresent(String s) {
        // Create a HashSet to store the two-character substrings
        HashSet<String> seenSubstrings = new HashSet<>();
        
        // Get the reverse of the string s
        String reversed = new StringBuilder(s).reverse().toString();
        
        // Iterate over the string to find all possible two-character substrings
        for (int i = 0; i < s.length() - 1; i++) {
            // Extract the substring of length 2 from the original string
            String substr = s.substring(i, i + 2);
            // Add to the HashSet
            seenSubstrings.add(substr);
        }
        
        // Check the reverse string for any substring that has already been seen
        for (int i = 0; i < reversed.length() - 1; i++) {
            String revSubstr = reversed.substring(i, i + 2);
            // If the reversed substring exists in the HashSet, return true
            if (seenSubstrings.contains(revSubstr)) {
                return true;
            }
        }
        
        // If no matching substring is found, return false
        return false;
    }
}
